make tests pass with enable_first_class_module() enabled.

zdevito opened the pull request:
Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Changed quantization so that it can find the parameters regardless of first-class mode or not.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21565 [WIP] make tests pass with enable_first_class_module() enabled.**
* #21515 Add WeakIValue, use in tracer.

Differential Revision: [D15729503](https://our.internmc.facebook.com/intern/diff/D15729503)

zdevito committed:
[WIP] make tests pass with enable_first_class_module() enabled.

zdevito committed:
Update on "[WIP] make tests pass with enable_first_class_module() enabled."

[WIP] make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "[WIP] make tests pass with enable_first_class_module() enabled."

[WIP] make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "[WIP] make tests pass with enable_first_class_module() enabled."

[WIP] make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "[WIP] make tests pass with enable_first_class_module() enabled."

[WIP] make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "[WIP] make tests pass with enable_first_class_module() enabled."

[WIP] make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "[WIP] make tests pass with enable_first_class_module() enabled."

[WIP] make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

make tests pass with enable_first_class_module() enabled.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

ZolotukhinM commented on test/test_jit.py:
```diff
@@ -10083,7 +10079,7 @@ def traced_fn(x):
 
     def test_call_script_mod_from_tracing_fn(self):
-        with self.disableEmitHook():
+        with self.assertRaisesRegex(RuntimeError, "must be registered as submodules"):
```
`FileCheck` call on line 10101 should probably be removed with this change.

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -48,59 +48,56 @@ Node* traverseToQuantNode(Node* dq) {
+  Value* definition;
+  IValue value;
+};
+static void gatherParamsFirstClass(script::Module& module, Value* module_value, std::vector<ParamValue>& params) {
```
Clang-format please 😛

ZolotukhinM commented on torch/csrc/jit/python_tracer.cpp:
```diff
@@ -79,7 +79,7 @@ std::shared_ptr<torch::jit::Graph> createGraphByTracing(
     }
     tracer::exit({toIValue(out)});
-    if (!script::getFirstClassMode()) {
+    if (true || !script::getFirstClassMode()) {
```
Is this some leftover?

zdevito replied:
yeah, Ill remove it

ZolotukhinM commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -3073,7 +3073,7 @@ void runCleanupPasses(std::shared_ptr<Graph>& to_clean, bool convert_ssa) {
   liftClosures(to_clean);
   inlineForkedClosures(to_clean);
-  if (!script::getFirstClassMode()) {
+  if (!script::getFirstClassMode() || true) {
```
And this.

ZolotukhinM commented on torch/csrc/jit/script/module.h:
```diff
@@ -99,6 +99,8 @@ struct TORCH_API Method {
     return *function_;
   }
 
+  std::pair<std::shared_ptr<Graph>, std::vector<at::Tensor>> _lowered_graph();
```
This needs some documentation.

ZolotukhinM reviewed (approved):
I'd prefer this to be splitted into several PRs, since some of the changes are completely independent - e.g. quantization and `requires_grad` changes. And you already proposed a good separation in the description. Apart from that, the changes look good with just some minor comments (inline).

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head

zdevito committed:
Update on "make tests pass with enable_first_class_module() enabled."

Grabbag of fixes required to make it possible to pass the test suite using first-class module execution:

* Fixed a bug where Objects were not considered requires_grad when one of their members might be. The default requires_grad now just checks contained types for requires grad, which should be more robust.
* Edited some tests to be less fragile by removing checks for implementation details and putting some tests on more standard interfaces.
* Provided `_lower_graph()` call that can be used by ONNX to still get the lowered exportable graph.

gh-metadata: pytorch pytorch 21565 gh/zdevito/60/head
